### PR TITLE
Reallow capture of memoryview arguments

### PR DESCRIPTION
**Pull request by da-woods**
It was actually OK in def functions. It looks very dodgy:

```
__Pyx_XDEC_MEMVIEW(closure->arg)
```

gets called twice and `INC` gets called once. However this is
actually OK since XDEC really means "clear" (this is why I
renamed it in master...)

Fixes https://github.com/cython/cython/issues/4798 (completely
I think).

Sorry for the mess with this!

### CODE REVIEW

**Comment by tristan957** [comment]
Hi @da-woods,

I assume this problem will go away when this PR is merged and released, but I am struggling to understand the error message on the latest 0.29.X release.

I have the following function

```
        def _iter(unsigned char [:]key_buf=None, unsigned char [:]value_buf=None):
            while True:
                key, val = self.read(key_buf=kbuf, value_buf=vbuf)
                if not self._eof:
                    yield key, val
                else:
                    return
```

```
Error compiling Cython file:
------------------------------------------------------------
...
            Iterator of key-value pairs.

        Raises:
            HseException: Underlying C function returned a non-zero value.
        """
        def _iter(unsigned char [:]key_buf=None, unsigned char [:]value_buf=None):
       ^
------------------------------------------------------------

subprojects/hse-python/hse3/hse.pyx:1336:8: Referring to a memoryview typed argument directly in a nested closure function is not supported in Cython 0.x. Either upgrade to Cython 3, or assign the argument to a local variable and use that in the nested function.
```

So I change the function to the following:

```
        def _iter(unsigned char [:]key_buf=None, unsigned char [:]value_buf=None):
            cdef unsigned char [:]kbuf = key_buf
            cdef unsigned char [:]vbuf = value_buf

            while True:
                key, val = self.read(key_buf=kbuf, value_buf=vbuf)
                if not self._eof:
                    yield key, val
                else:
                    return
```

And I get the same error message. I don't understand what I am doing wrong. I have looked at what the scipy people did to fix their issues, and I don't feel like mine are any different.

**Comment by da-woods** [comment]
The assignment
```
cdef unsigned char [:]kbuf = key_buf
cdef unsigned char [:]vbuf = value_buf
```
needs to occur *outside* `_iter`. But yes, it should go away when this PR is merged

**Comment by scoder** [comment]
In the case of scipy, they used an inner function, which means that the call directly from the outer function is optimised into a direct C call. Not sure if that makes a difference here.

**Comment by tristan957** [comment]
For completeness in case others stumble across the same mistakes I made.

```
diff --git a/hse3/hse.in.pyx b/hse3/hse.in.pyx
index d1f9939..0f88ae0 100644
--- a/hse3/hse.in.pyx
+++ b/hse3/hse.in.pyx
@@ -857,15 +857,18 @@ cdef class KvsCursor:
         """
         @SUB@ hse.KvsCursor.items
         """
-        def _iter(unsigned char [:]key_buf=None, unsigned char [:]value_buf=None):
+        cdef unsigned char [:]kbuf = key_buf
+        cdef unsigned char [:]vbuf = value_buf
+
+        def _iter():
             while True:
-                key, val = self.read(key_buf=key_buf, value_buf=value_buf)
+                key, val = self.read(key_buf=kbuf, value_buf=vbuf)
                 if not self._eof:
                     yield key, val
                 else:
                     return
 
-        return _iter(key_buf=key_buf, value_buf=value_buf)
+        return _iter()
 
     def update_view(self) -> None:
         """
```

Got my code to compile. Thanks!

File: tests/memoryview/memslice.pyx [hunk 1]
```diff
@@ -2568,4 +2587,3 @@ def test_arg_in_closure_cdef(a):
     released A
     """
     return arg_in_closure_cdef(a)
-
```
Comment (scoder) [inline tests/memoryview/memslice.pyx:2571]: Probably not intended.

**Comment by xkszltl** [comment]
Should I expect 0.29.32 to be released very soon, or should I cherry-pick this instead?

**Comment by da-woods** [comment]
> Should I expect 0.29.32 to be released very soon, or should I cherry-pick this instead?

Possibly cherry-pick it yourself (or cherry-pick the change in Scipy, or downgrade to Cython 0.29.30). Other people's definition of soon may not be the same as yours

**Comment by scoder** [comment]
Thanks. I'll try to get 0.29.32 out quickly then.

**Comment by xkszltl** [comment]
> > Should I expect 0.29.32 to be released very soon, or should I cherry-pick this instead?
> 
> Possibly cherry-pick it yourself (or cherry-pick the change in Scipy, or downgrade to Cython 0.29.30). Other people's definition of soon may not be the same as yours

Good point

**Comment by da-woods** [comment]
0.29.32 is now out so you shouldn't need to cherry-pick anything. It contains another memoryview-related bug fix which you quite likely also want

**Comment by xkszltl** [comment]
Yes, already integrating~
Thanks for the quick action this!

**Pull request merged**